src/lsm/tree.zig

```diff
@@ -228,32 +235,119 @@ pub fn TreeType(comptime Table: type) type {
                 if (tree.table_mutable.get(key) orelse
                     tree.value_cache.?.getKeyPtr(tombstone_from_key(key))) |value|
                 {
-                    callback(unwrap_tombstone(value));
+                    callback(context, unwrap_tombstone(value));
                     return;
                 }
             }
 
-            // Hash the key to the fingerprint only once and reuse for all bloom filter checks.
             if (!tree.table_immutable.free and tree.table_immutable.snapshot_min < snapshot) {
                 if (tree.table_immutable.get(key)) |value| {
-                    callback(unwrap_tombstone(value));
+                    callback(context, unwrap_tombstone(value));
                     return;
                 }
             }
 
-            var it = tree.manifest.lookup(snapshot, key);
-            if (it.next()) |info| {
-                assert(info.visible(snapshot));
-                assert(compare_keys(key, info.key_min) != .lt);
-                assert(compare_keys(key, info.key_max) != .gt);
+            // Hash the key to the fingerprint only once and reuse for all bloom filter checks.
+            const fingerprint = bloom_filter.Fingerprint.create(mem.asBytes(&key));
+
+            context.* = .{
+                .tree = tree,
+                .completion = undefined,
```

Comment:
I don't have any strong preference either way, this is just the style we've been using everywhere else for this pattern (journal, grid, etc.).

I think I started doing it this way because I found it a bit more obvious what was going on with completion at the usage site.